internal/allocator/allocator.go

```diff
@@ -381,3 +372,168 @@ func ipConfusesBuggyFirmwares(ip net.IP) bool {
 	}
 	return ip[3] == 0 || ip[3] == 255
 }
+
+// Dual-stack.
+
+// AssignDual assigns the requested ip's to svc, if the assignment is
+// permissible by sharingKey and backendKey.
+func (a *Allocator) AssignDual(svc string, ip, ip2 net.IP, ports []Port, sharingKey, backendKey string) error {
+	if ipIsIPv6(ip) == ipIsIPv6(ip2) {
+		return fmt.Errorf("%q %q is has the same family", ip, ip2)
+	}
+	pool := poolFor(a.pools, ip)
+	if pool == "" {
+		return fmt.Errorf("%q is not allowed in config", ip)
+	}
+	if poolFor(a.pools, ip2) != pool {
+		return fmt.Errorf("%q %q different pools", ip, ip2)
+	}
+
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+
+	if err := a.checkSharing(svc, ip.String(), ports, sk); err != nil {
+		return err
+	}
+	if err := a.checkSharing(svc, ip2.String(), ports, sk); err != nil {
+		return err
+	}
+
+	// Either the IP is entirely unused, or the requested use is
+	// compatible with existing uses. Assign! But unassign first, in
+	// case we're mutating an existing service (see the "already have
+	// an allocation" block above). Unassigning is idempotent, so it's
+	// unconditionally safe to do.
+	alloc := &alloc{
+		pool:  pool,
+		ip:    ip,
+		ip2:   ip2,
+		ports: make([]Port, len(ports)),
+		key:   *sk,
+	}
+	for i, port := range ports {
+		port := port
```

Comment:
why u need another port variable here ?

Reply:
Seems like a misstake. Will fix it. (strange that it even compiles)

Reply:
But it it a copy/paste thing. The exact same code is in the Assign() function further up. I'll fix that too.